core/src/main/java/org/elasticsearch/ingest/PipelineStore.java

```diff
@@ -173,280 +157,91 @@ public void delete(DeletePipelineRequest request, ActionListener<DeleteResponse>
-            client.admin().indices().create(createIndexRequest, new ActionListener<CreateIndexResponse>() {
-                @Override
-                public void onResponse(CreateIndexResponse createIndexResponse) {
-                    innerPut(request, listener);
-                }
+            @Override
+            protected WritePipelineResponse newResponse(boolean acknowledged) {
+                return new WritePipelineResponse(acknowledged);
+            }
 
-                @Override
-                public void onFailure(Throwable e) {
-                    listener.onFailure(e);
-                }
-            });
-        }
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                return innerPut(request, currentState);
+            }
+        });
     }
 
-    private void innerPut(PutPipelineRequest request, ActionListener<IndexResponse> listener) {
-        IndexRequest indexRequest = new IndexRequest(request);
-        indexRequest.index(PipelineStore.INDEX);
-        indexRequest.type(PipelineStore.TYPE);
-        indexRequest.id(request.id());
-        indexRequest.source(request.source());
-        indexRequest.refresh(true);
-        client.index(indexRequest, handleWriteResponseAndReloadPipelines(listener));
+    ClusterState innerPut(PutPipelineRequest request, ClusterState currentState) {
+        IngestMetadata currentIngestMetadata = currentState.metaData().custom(IngestMetadata.TYPE);
+        Map<String, PipelineConfiguration> pipelines;
+        if (currentIngestMetadata != null) {
+            pipelines = new HashMap<>(currentIngestMetadata.getPipelines());
+        } else {
+            pipelines = new HashMap<>();
+        }
+
+        pipelines.put(request.id(), new PipelineConfiguration(request.id(), request.source()));
+        ClusterState.Builder newState = ClusterState.builder(currentState);
+        newState.metaData(MetaData.builder(currentState.getMetaData())
+            .putCustom(IngestMetadata.TYPE, new IngestMetadata(pipelines))
+            .build());
+        return newState.build();
     }
 
     /**
      * Returns the pipeline by the specified id
      */
     public Pipeline get(String id) {
-        ensureReady();
-
-        PipelineDefinition ref = pipelines.get(id);
-        if (ref != null) {
-            return ref.getPipeline();
-        } else {
-            return null;
-        }
+        return pipelines.get(id);
     }
 
     public Map<String, Processor.Factory> getProcessorFactoryRegistry() {
         return processorFactoryRegistry;
     }
 
-    public List<PipelineDefinition> getReference(String... ids) {
-        ensureReady();
+    /**
+     * @return a pipeline as maps of maps by id. Multiple pipelines will be returned if multiple ids are provided.
+     */
+    // Returning PipelineConfiguration instead of Pipeline, because Pipeline and Processor interface don't
+    // know how to serialize themselves.
```

Comment:
right, this is just a note about serialization. In the top of the file I mention the reasons why pipelines can't be part of cluster state.